domain/src/main/java/co/nimblehq/domain/ResponseMapping.kt

```diff
@@ -0,0 +1,47 @@
+package co.nimblehq.domain
+
+import co.nimblehq.data.service.error.JsonApiException
+import co.nimblehq.data.service.error.NoConnectivityException
+import co.nimblehq.data.service.error.UnknownException
+import co.nimblehq.data.service.providers.MoshiBuilderProvider
+import co.nimblehq.data.service.response.ErrorResponse
+import com.squareup.moshi.JsonAdapter
+import io.reactivex.Single
+import retrofit2.Response
+import java.io.InterruptedIOException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+
+fun <T> Single<Response<T>>.transform(): Single<T> = flatMap {
+    val body = it.body()
+    if (it.isSuccessful && body != null) {
+        Single.just(body)
+    } else {
+        Single.error(mapError(it))
+    }
+}.onErrorResumeNext {
+    when (it) {
+        is SocketTimeoutException -> Single.error(it)
+        is UnknownHostException,
+        is InterruptedIOException -> Single.error(NoConnectivityException)
+        else -> Single.error(it)
+    }
+}
+
+private fun <T> mapError(response: Response<T>?): Exception {
+    val errorBody = response?.errorBody()?.string()
+    val errorResponse = parseErrorResponse(errorBody)
+
+    return if (errorResponse != null && response != null) JsonApiException(errorResponse, response)
+    else UnknownException
+}
+
+fun parseErrorResponse(source: String?): ErrorResponse? {
+    return try {
+        val moshi = MoshiBuilderProvider.moshiBuilder.build()
+        val adapter: JsonAdapter<ErrorResponse> = moshi.adapter(ErrorResponse::class.java)
```

Comment:
I think we could move this to a separate provider class rather than put it here. What do you think? 🤔

Reply:
we had `MoshiBuilderProvider` to build moshi instance already. I can't get your idea what to improve here 💭

Reply:
Oh. The adapter needs Moshi to init. Could we remove `: JsonAdapter<ErrorResponse>` btw?

Reply:
@hoangnguyen92dn done ✅